cdap-docs/developer-manual/source/metadata/programmatic-metadata.rst

```diff
@@ -0,0 +1,85 @@
+.. meta::
+    :author: Cask Data, Inc.
+    :copyright: Copyright © 2018 Cask Data, Inc.
+
+.. _programmatic-metadata:
+
+==================================
+Metadata from Programs and Plugins
+==================================
+
+.. highlight:: java
+
+Metadata can be accessed from *Programs* and *Plugins*. Currently, reading metadata is not supported while CDAP is running in *cloud mode* but metadata can be added, updated or deleted in all modes.
+
+.. _metadata-programs:
+
+Programs
+========
+
+Metadata can be accessed from *MapReduce*, *Spark*, *Workers* and *Service* through their individual program context.
+Metadata operations are not supported at individual tasks and should be performed only during the
+initialization phase of the program.
+
+For example metadata can be read in *MapReduce* programs in the following way::
+
+  @Override
+  public void initialize() throws Exception {
+    MapReduceContext context = getContext();
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A program can also emit metadata properties and tags using the context.
+For example a dataset can be tagged from a *MapReduce* program in the following way::
+
+  @Override
+  public void initialize() throws Exception {
+    MapReduceContext context = getContext();
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // add a tag
+    context.addTags(entity, "someTag");
+  }
+
+.. _metadata-plugins:
+
+Plugins
+=======
+A *Plugin* can also access metadata through the context provided to it in ``prepareRun`` stage. Like *programs*, metadata operations are not supported at individual task level.
+
+Metadata can be read in a plugin in the following way::
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    context.setInput(Input.ofDataset(config.tableName));
+
+    // construct the metadata entity
+    MetadataEntity entity = MetadataEntity.builder().append(MetadataEntity.NAMESPACE, "myNamespace")
+      .appendAsType(MetadataEntity.DATASET, "myDataset").build();
+
+    // get the metadata
+    Map<MetadataScope, Metadata> metadata = context.getMetadata(entity);
+  }
+
+A plugin can also emit metadata. For example a plugin can tag a dataset in the following way::
```

Comment:
similar change about annotating metadata to an entity here

Reply:
Done.